### PR TITLE
core: replace ShardChunkHeader::version_range by valid_for

### PR DESCRIPTION
**Pull request by mina86**
Replace ShardChunkHeader::version_range and
PartialEncodedChunk::version_range methods with valid_for method which
is simpler interface for the call site and which allows us to get rid
of custom ProtocolVersionRange type.

Note that, yes, PartialEncodedChunk::valid_for is actually unused.
I’m leaving it in though.  Perhaps it’s worth keeping the record of
when each version of the PartialEncodedChunk was introduced?

### CODE REVIEW

**Comment by matklad** [comment]
This would be a great trade if only `std::ops::Range` wasn't such a misshapen type: https://ridiculousfish.com/blog/posts/least-favorite-rust-type.html (no Copy, no strong invariants). Still, I think it is worth it on balance.

**Review by matklad — APPROVED** [review]

File: core/primitives/src/sharding.rs [hunk 1]
```diff
@@ -399,17 +397,16 @@ impl ShardChunkHeader {
...
     }
 
-    pub fn version_range(&self) -> ProtocolVersionRange {
-        let block_header_v3_version = ProtocolFeature::BlockHeaderV3.protocol_version();
+    /// Returns the range of `ProtocolVersion`s where this variant of the
+    /// message is accepted.  For the newest variant the upper limit is the
+    /// maximum protocol version.
+    pub fn version_range(&self) -> std::ops::Range<ProtocolVersion> {
+        const BLOCK_HEADER_V3_VERSION: ProtocolVersion =
+            ProtocolFeature::BlockHeaderV3.protocol_version();
```
Comment (matklad) [inline core/primitives/src/sharding.rs:405]: Wait, why do we need a `const` here? Just `let` would do. I don't think it's worth using `const` for local variables just to emphasize constness: 

* the optimizer is smart enough to constant-fold (if it wasn't, the whole iterator infrastructure would be horribly slow)
* SCREAMING_SNAKE_CASE looks jarring
* `const` is an item, it has very different name resolution rules from `let`

That being said, everything else here is CAPS already, so perhaps this makes code more uniform? No strong opinion either way.

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -1556,7 +1556,7 @@ impl ShardsManager {
 
         // 2. check protocol version
         let protocol_version = self.runtime_adapter.get_epoch_protocol_version(&epoch_id)?;
-        if !header.version_range().contains(protocol_version) {
```
Comment (matklad) [inline chain/chunks/src/lib.rs:1559]: Looking at the call-site, perhaps we don't need a range here at all and just `header.valid_for(protocol_version)` ?

File: core/primitives/src/sharding.rs (hunk 1 above)
Comment (mina86) [reply core/primitives/src/sharding.rs:403]: > That being said, everything else here is CAPS already, so perhaps this makes code more uniform?

Yeah, that’s pretty much my reasoning.  It IMO looks more aesthetic.

File: chain/chunks/src/lib.rs [hunk 3]
```diff
@@ -1556,14 +1556,13 @@ impl ShardsManager {
...
-        if !header.version_range().contains(&protocol_version) {
-            return if epoch_id_confirmed {
-                Err(Error::InvalidChunkHeader)
-            } else {
-                Err(DBNotFoundErr(format!("block {:?}", header.prev_block_hash())).into())
-            };
+        if header.valid_for(protocol_version) {
+            Ok(())
+        } else if epoch_id_confirmed {
+            Err(Error::InvalidChunkHeader)
```
[the remark this answers is not in the document]
Comment (mina86) [reply chain/chunks/src/lib.rs:1564]: Refactored the if a bit.

**Review by matklad — APPROVED** [review]

**Pull request merged**